### PR TITLE
dpu

### PR DESCRIPTION
**Pull request by samsja**

### CODE REVIEW

File: src/zeroband/global_ddp.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+class GlobalDDP:
+    """
+    This class implements DDP over internet. It
+
+    :Args:
+        model: The model to be trained
+        elastic_device_mesh: The elastic device mesh to be used
+        dpu: Whether to use delayed parameter updates
+
```
Comment (Jackmin801) [inline src/zeroband/global_ddp.py]: Args have changed

File: src/zeroband/global_ddp.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    Example usage:
+
+    ```
+    global_ddp = GlobalDDP(model, elastic_device_mesh)
```
Comment (Jackmin801) [inline src/zeroband/global_ddp.py]: Example changed

File: src/zeroband/global_ddp.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+    for step in range(num_steps):
+        for micro_bs in range(num_micro_bs):
+            optimizer.zero_grad()
+            loss = model(batch)
+            loss.backward()
+
+        global_ddp.all_reduce()
+        optimizer.step()
+        diloco.step(model)
```
Comment (Jackmin801) [inline src/zeroband/global_ddp.py]: The optimizer zero grad is in the wrong place? This would make only the last micro batch have grad

File: src/zeroband/global_ddp.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    ):
+        self.elastic_device_mesh = elastic_device_mesh
+        self.config = config
+
+        self.world_info = get_world_info()
+        self._logger = get_logger()
+
+        self.model = model
+
+        self._staling_grad_work: list[AllReduceGradWork] | None = None
```
Comment (Jackmin801) [inline src/zeroband/global_ddp.py]: staling -> stalling

File: src/zeroband/global_ddp.py (hunk 3 above)
Comment (samsja) [reply src/zeroband/global_ddp.py]: lol good catch

File: src/zeroband/global_ddp.py (hunk 4 above)
Comment (samsja) [reply src/zeroband/global_ddp.py]: staline gradient ?

Comment (Jackmin801) [reply src/zeroband/global_ddp.py]: stalingrad

Comment (samsja) [reply src/zeroband/global_ddp.py]: hahahahahah
